scalus-core/shared/src/test/scala/scalus/builtin/BuiltinArrayTest.scala

```diff
@@ -0,0 +1,132 @@
+        val result = term.evaluateDebug
+        result match {
+            case Result.Success(evaled, _, _, _) =>
+                assert(evaled == Term.Const(scalus.uplc.Constant.Integer(2)))
+            case Result.Failure(err, _, _, _) =>
+                fail(s"Expected success but got failure: $err")
+        }
+    }
+
+    test("empty array has length 0") {
+        val sir = compile {
+            val emptyList: BuiltinList[Data] = mkNilData()
+            val array = listToArray(emptyList)
+            lengthOfArray(array)
+        }
+        val term = sir.toUplc()
+        val result = term.evaluateDebug
+        result match {
+            case Result.Success(evaled, _, _, _) =>
+                assert(evaled == Term.Const(scalus.uplc.Constant.Integer(0)))
+            case Result.Failure(err, _, _, _) =>
+                fail(s"Expected success but got failure: $err")
+        }
+    }
+
+    test("indexArray with out of bounds index fails") {
+        val sir = compile {
+            val list = mkCons(iData(1), mkNilData())
+            val array = listToArray(list)
+            indexArray(array, BigInt(5)) // Out of bounds
+        }
+        val term = sir.toUplc()
+        val result = term.evaluateDebug
+        assert(result.isFailure, s"Expected failure but got: $result")
+    }
+
+    test("indexArray with first element (index 0)") {
+        val sir = compile {
+            val list = mkCons(iData(100), mkCons(iData(200), mkNilData()))
+            val array = listToArray(list)
+            indexArray(array, BigInt(0))
+        }
+        val term = sir.toUplc()
+        val result = term.evaluateDebug
+        result match {
+            case Result.Success(evaled, _, _, _) =>
+                assert(evaled == Term.Const(scalus.uplc.Constant.Data(Data.I(100))))
+            case Result.Failure(err, _, _, _) =>
+                fail(s"Expected success but got failure: $err")
+        }
+    }
+
+    test("indexArray with last element") {
+        val sir = compile {
+            val list = mkCons(iData(1), mkCons(iData(2), mkCons(iData(3), mkNilData())))
+            val array = listToArray(list)
+            indexArray(array, BigInt(2))
+        }
+        val term = sir.toUplc()
+        val result = term.evaluateDebug
+        result match {
+            case Result.Success(evaled, _, _, _) =>
+                assert(evaled == Term.Const(scalus.uplc.Constant.Data(Data.I(3))))
+            case Result.Failure(err, _, _, _) =>
+                fail(s"Expected success but got failure: $err")
+        }
+    }
+
+    test("indexArray with negative index fails") {
+        val sir = compile {
+            val list = mkCons(iData(1), mkCons(iData(2), mkNilData()))
+            val array = listToArray(list)
+            indexArray(array, BigInt(-1))
+        }
+        val term = sir.toUplc()
+        val result = term.evaluateDebug
+        assert(result.isFailure, s"Expected failure but got: $result")
+    }
+}
```

Comment:
If you represent BuitonArray as case class, than user
- can match on it
- can select field
- can construct field